Initial implementation of the `ai` extension

elprans opened the pull request:
This adds the `ai` extension to EdgeDB, containing the following
functionality:

1. The new object-level `ext::ai::index` (similar to `fts::index`) that
   automatically generates and indexes embeddings from the given index
   expression.

2. A basic RAG interface via the `/ai/rag` HTTP endpoint that takes
   a query selecting from an `ai::index`-indexed type and uses it as
   context for a text generation question-answer completion.

The support for the above is implemented here for OpenAI, Mistral and
Anthropic, mostly to demonstrate multi-provider support.  Model and
provider metadata is implemented as annotated abstract types in `ext::ai`,
and the intent is that users can define models in their schemas by
extending from the appropriate base type in `ext::ai`.

To test this out:

1. Apply the following schema:

   ```edgeql
   using extension ai;

   module default {
     type Astronomy {
       content: str;
       deferred index ext::ai::index(embedding_model := 'text-embedding-3-small')
         on (.content);
     }
   };
   ```

2. Configure the OpenAI provider:

   ```edgeql
   configure current database
   insert ext::ai::OpenAIProviderConfig {
     secret := 'sk-....',
   };
   ```

3. Insert some data:

   ```edgeql
   insert Astronomy {
     content := 'Skies on Mars are red'
   };
   insert Astronomy {
     content := 'Skies on Earth are blue'
   };
   ```

4. Test RAG:

   ```shell
   $ curl --json '{
       "query": "What color is the sky on Mars?",
       "model": "gpt-4-turbo-preview",
       "context": {"query":"Astronomy"}
     }' http://127.0.0.1:5656/branch/main/ai/rag
   
   {"response": "The sky on Mars is red."}
   ```

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+        ),
+    )
+
+    my_options = dataclasses.replace(options, singletons=frozenset())
```
Why clearing singletons?

elprans replied:
What we are doing here is:

```
SELECT (ObjType.id, ObjType."index_expr_of_ext_ai_index")
```

We want all objects, not a single one, but since we're recompiling the `index.expr` we'd better use the options for it.

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+# --- pgvector ---
+
+
+def _pg_create_ai_embeddings(
+    index: s_indexes.Index,
+    options: qlcompiler.CompilerOptions,
+    predicate_src: Optional[str],
+    sql_kwarg_exprs: dict[str, str],
+    schema: s_schema.Schema,
+) -> dbops.Command:
```
Some more comments would be good, to make it clear what all gets created here

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+    return ops
+
+
+def _pg_create_trigger(
+    schema: s_schema.Schema,
+    index: s_indexes.Index,
+    table_name: tuple[str, str],
+    dep_cols: list[str],
+) -> dbops.Command:
+    ops = dbops.CommandGroup()
```
Mention what the trigger does?

msullivan replied:
And maybe add a comment that we might want to share some machinery with FTS eventually

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+    return ops
+
+
+def pg_rebuild_all_pending_ebeddings_views(
```
s/ebeddings/embeddings/

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+    return ops
+
+
+def pg_drop_all_pending_ebeddings_views(
```
s/ebeddings/embeddings/

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+        sn.QualName("ext::ai", "embedding_model_output_dimensions"),
+        checked.CheckedList[int],
+    )
+    return list(sorted(dimensions))[-1]
```
`max(dimensions)`?

msullivan commented on edb/pgsql/delta_ext_ai.py:
```diff
@@ -0,0 +1,678 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2008-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Could you document in the source what all the columns added, views added, etc are?

msullivan reviewed (commented):
A first batch of comments on the schema and index commits.

Mostly they are requests for more documentation

msullivan commented on edb/pgsql/compiler/relgen.py:
```diff
@@ -3898,6 +3899,89 @@ def process_set_as_fts_search(
+    if target_index_metadata is None:
+        raise AssertionError(
+            "missing expected index metadata in FunctionCall.extras")
+    index_id = target_index_metadata.get("id")
+    if index_id is None:
+        raise AssertionError(
+            "missing expected index metadata in FunctionCall.extras")
+    dimensions = target_index_metadata.get("dimensions")
+    if dimensions is None:
+        raise AssertionError(
+            "missing expected index metadata in FunctionCall.extras")
+    df = target_index_metadata.get("distance_function")
+    if index_id is None:
+        raise AssertionError(
+            "missing expected index metadata in FunctionCall.extras")
```
Maybe extract with a single `match`?

msullivan commented on edb/pgsql/compiler/relgen.py:
```diff
@@ -3898,6 +3899,89 @@ def process_set_as_fts_search(
+    call: irast.Call,
+    obj_id: irast.PathId,
+    args_pg: list[pgast.BaseExpr],
+    _ctx: context.CompilerContextLevel,
+    newctx: context.CompilerContextLevel,
+    _inner_ctx: context.CompilerContextLevel,
```
This one doesn't need to be addressed yet since it's really an FTS thing, but having three different contexts is quite complex and needs to be explained and given likely better names. (I think it's because the FTS stuff is picky about where stuff goes?)

msullivan replied:
@aljazerzen ^

msullivan commented:
This is going to need tests. Testing the integrations themselves might not be plausible, but there is a ton of other code that we want to test.

Possible approaches:
1. Create a dummy provider that we use for tests
2. Stand up a local HTTP server that mocks one of the existing providers

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,391 @@
+                            processed = await _ext_ai_index_builder_work(
+                                db, pgconn, models)
+                        finally:
+                            if processed == 0:
+                                await _ext_ai_unlock(pgconn)
```
The locking/unlocking seems wrong. We'll potentially acquire the lock multiple times (which is permitted) but only unlock it once?

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -324,6 +324,102 @@ async def _generate_embeddings(
+        response.close_connection = True
+        return
+    if request.content_type != b"application/json":
+        response.body = b"Expected applicatoin/json input"
```
s/applicatoin/application/

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -324,6 +324,19 @@ async def _generate_embeddings(
     raise RuntimeError(f"unsupported model provider: {provider.name}")
 
 
+async def _start_chat(
```
RAG API docs need to be reflected somewhere more than a commit message

elprans replied:
For sure.  The commit message is a starting point.

elprans commented on edb/lib/ext/ai.edgeql:
```diff
@@ -0,0 +1,213 @@
+        using sql expression;
+    };
+
+    create scalar type ext::ai::ChatParticipantRole
+        extending enum<System, User, Assistant, Tool>;
```
For now this is the exhaustive list for common APIs.

elprans commented on edb/pgsql/delta.py:
```diff
@@ -3707,12 +3720,12 @@ def create_index(
         return dbops.CreateIndex(pg_index)
 
-    def _create_begin(
+    def _create_innards(
```
Because we need the transplanted-from-model annotations to have been applied to the index and those are "innards".

elprans commented on edb/schema/indexes.py:
```diff
@@ -1244,6 +1265,80 @@ def validate_object(
+                    span=self.span,
+                )
+
+            model_stype = next(iter(models.values()))
+            model_stype_vn = model_stype.get_verbosename(schema)
+            model_annos = model_stype.get_annotations(schema)
+            my_name = self.scls.get_name(schema)
+            for model_anno in model_annos.objects(schema):
+                anno_name = model_anno.get_shortname(schema)
+                if anno_name.module != "ext::ai":
+                    continue
+                value = model_anno.get_value(schema)
+                if value is None or value == "<must override>":
+                    raise errors.SchemaDefinitionError(
+                        f"{model_stype_vn!r} is missing a value for the "
+                        f"{anno_name!r} annotation"
+                    )
+                anno_sname = sn.get_specialized_name(
+                    anno_name,
+                    str(my_name),
+                )
+                schema1 = model_anno.update(
+                    schema,
+                    {
+                        "name": sn.QualName(my_name.module, anno_sname),
+                        "subject": self.scls,
+                    },
+                )
+                anno_copy = schema1.get(
+                    model_anno.get_name(schema1),
+                    type=s_anno.AnnotationValue,
+                )
+                anno_cmd = anno_copy.as_create_delta(
+                    schema1, so.ComparisonContext())
+                anno_cmd.discard_attribute("bases")
+                anno_cmd.discard_attribute("ancestors")
+                self.add(anno_cmd)
```
Yeah, we are essentially copying the relevant annotations from the declaration of the model specified in the `embedding_model` index kwarg.  This is needed because looking those annotations up from the model at run time would be very inefficient.

Bases are cleared because we don't want to "inherit" the annotations from the abstract base model's definition.

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -363,6 +378,264 @@ async def handle_request(
+        if not isinstance(context, dict):
+            raise TypeError(
+                '"context" value in request is not a valid JSON object')
+
+        ctx_query = context.get("query")
+        ctx_variables = context.get("variables")
+        ctx_globals = context.get("globals")
+        ctx_max_obj_count = context.get("max_object_count")
+
+        if not ctx_query:
+            raise TypeError(
+                'missing required "query" in request "context" object')
+
+        if ctx_variables is not None and not isinstance(ctx_variables, dict):
+            raise TypeError('"variables" must be a JSON object')
+
+        if ctx_globals is not None and not isinstance(ctx_globals, dict):
+            raise TypeError('"globals" must be a JSON object')
+
+        model = body.get('model')
+        if not model:
+            raise TypeError(
+                'missing required "model" in request')
+
+        query = body.get('query')
+        if not query:
+            raise TypeError(
+                'missing required "query" in request')
+
+        stream = body.get('stream')
+        if stream is None:
+            stream = False
+        elif not isinstance(stream, bool):
+            raise TypeError('"stream" must be a boolean')
+
+        if ctx_max_obj_count is None:
+            ctx_max_obj_count = 5
+        elif not isinstance(ctx_max_obj_count, int) or ctx_max_obj_count <= 0:
+            raise TypeError(
+                '"context.max_object_count" must be an postitive integer')
```
Not a comment for now, but I think we're going to want a lightweight HTTP API framework at some point...

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -363,6 +378,264 @@ async def handle_request(
+        ctx_globals = context.get("globals")
+        ctx_max_obj_count = context.get("max_object_count")
+
+        if not ctx_query:
+            raise TypeError(
+                'missing required "query" in request "context" object')
```
Should we make sure that it parses as a standalone fragment?

elprans replied:
why?  If it parses while wrapped then that's good enough, no?

msullivan replied:
I worry that some sort of injection might be possible.

msullivan replied:
Another thing we need to handle is comments in the query.

For that it probably suffices to append a newline after?

elprans replied:
The API here isn't intended to be used by untrusted parties, it is similar in that regard to edgeql+http or graphql.  We also explicitly disable all capabilities, including mutation.

msullivan replied:
Maybe my answer is just "it squicks me out", then

msullivan replied:
I think there are queries with unbalanced parens that would work

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -333,9 +365,158 @@ async def _start_chat(
+                "messages": messages,
+                "stream": True,
+            }
+        ) as event_source:
+            async for sse in event_source.aiter_sse():
+                if not response.sent:
+                    response.status = http.HTTPStatus.OK
+                    response.content_type = b'text/event-stream'
+                    response.close_connection = False
+                    response.custom_headers["Cache-Control"] = "no-cache"
+                    protocol.write(request, response)
+
+                if sse.event != "message":
+                    continue
+
+                if sse.data == "[DONE]":
+                    event = (
+                        b'event: message_stop\n'
+                        + b'data: {"type": "message_stop"}\n\n'
+                    )
+                    protocol.write_raw(event)
+                    continue
```
Is the goal here that we a normalizing to an anthropic-like API?

msullivan commented:
Really impressive work.

I might still have some more questions about the integrations.

jaclarke commented on edb/lib/ext/ai.edgeql:
```diff
@@ -0,0 +1,405 @@
+            create annotation std::description := "Provider API URL.";
+        };
+
+        create property client_id: std::str {
```
Should this also be required? Without the client id, the server logs the error below when it has objects to index:
```
------------------------- Exception occurred: unhandled errors in a TaskGroup (1 sub-exception) --------------------------

1. ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)

------------------------------------------------------- Traceback --------------------------------------------------------

    /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 107, in _ext_ai_index_builder_controller_loop
        > processed = await _ext_ai_index_builder_work(
    /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 234, in _ext_ai_index_builder_work
        > async with asyncio.TaskGroup() as g:
    /opt/homebrew/Cellar/python@3.12/3.12.2_1/Frameworks/Python.framework/Versions/3.12/lib/python3.12/asyncio/taskgroups.py, line 145, in __aexit__
        > raise me from None

ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
--------------------------------------------------- Grouped exceptions ---------------------------------------------------


    -------------------------- Exception occurred: 'NoneType' object has no attribute 'encode' ---------------------------

    1. AttributeError: 'NoneType' object has no attribute 'encode'

    ----------------------------------------------------- Traceback ------------------------------------------------------

        /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 329, in _generate_embeddings
            > return await _generate_openai_embeddings(provider, model_name, inputs)
        /Users/james/dev/edgedb/edb/server/protocol/ai_ext.py, line 341, in _generate_openai_embeddings
            > client = httpx.AsyncClient(
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_client.py, line 1371, in __init__
            > super().__init__(
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_client.py, line 182, in __init__
            > self.headers = Headers(headers)
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_models.py, line 75, in __init__
            > normalize_header_value(v, encoding),
        /Users/james/dev/edgedb-dev/lib/python3.12/site-packages/httpx/_utils.py, line 53, in normalize_header_value
            > return value.encode(encoding or "ascii")

    AttributeError: 'NoneType' object has no attribute 'encode'


ERROR 18430 _localdev 2024-04-12T14:11:40.496 edb.server.ai_ext: caught error in ext::ai::index builder on database 'main'
```
and requests to the rag endpoint return the same error:
```
curl --json '{
    "query": "What color is the sky on Mars?",
    "model": "gpt-4-turbo-preview",
    "context": {"query":"Astronomy"}
  }' http://localhost:5656/branch/main/ai/rag
AttributeError: 'NoneType' object has no attribute 'encode'%    
```

jaclarke replied:
Also when I removed the provider config to insert a new version with the client id, the server logged this error:
```
ERROR 21062 _localdev 2024-04-12T14:21:01.151 edb.server.ai_ext: ext::ai::index builder on database 'main': provider 'builtin::openai' has not been configured
```

elprans replied:
OpenAI doesn't require it, so I made it optional.

elprans commented:
Pushed a bunch of fixup commits.  I realized that I initially misunderstood how embedding shortening works, so I reworked the model metadata around that and also removed creation of discrete subvector columns for shortening-capable embeddings as it should be possible to create expression indexes on subvectors directly (with the yet-unreleased `subvector()` in pgvector).

elprans commented:
Oh, and I fixed handling of `text-embedding-3-large` by truncating its output to 2000 (pgvector maximum).

msullivan commented on edb/server/protocol/ai_ext.py:
```diff
@@ -1235,6 +1235,23 @@ async def _generate_embeddings_for_type(
                         LIMIT
                             1
                     ).0,
+                    model_embedding_dimensions := <int64>(
+                        SELECT
+                            (.annotations@value, .annotations.name)
+                        FILTER
+                            .1 =
+                            "ext::ai::embedding_model_max_output_dimensions"
+                        LIMIT
+                            1
+                    ).0,
+                    index_embedding_dimensions := <int64>(
+                        SELECT
+                            (.annotations@value, .annotations.name)
+                        FILTER
+                            .1 = "ext::ai::embedding_dimensions"
+                        LIMIT
+                            1
+                    ).0,
```
This change is in a fixup to `/ai/rag` but seems like it is for embeddings?

msullivan commented:
Looking good, I think!

1st1 commented on edb/lib/ext/ai.edgeql:
```diff
@@ -0,0 +1,420 @@
+    create abstract type ext::ai::ProviderConfig extending cfg::ConfigObject {
+        create required property name: std::str {
+            set readonly := true;
+            create constraint exclusive;
```
Please add a `std::description` annotation here too

1st1 commented on edb/lib/ext/ai.edgeql:
```diff
@@ -0,0 +1,420 @@
+            ext::ai::model_provider := "builtin::anthropic";
+        alter annotation
+            ext::ai::text_gen_model_context_window := "200000";
+    };
```
I guess the only gripe I have about this is that it's sad we're polluting our namespace with all these model names, as they will eventually go obsolete.

1st1 commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,1305 @@
+                                    db, pgconn, models)
+                            finally:
+                                if processed == 0:
+                                    await _ext_ai_unlock(pgconn)
```
I don't trust awaits in finally (and I don't remember enough to say confidently that ext_ai_unlock is 100% guaranteed to run and not leave `pgcon` still holding the lock with this coroutine cancelled). The main question is if `tenant.release_pgcon(dbname, pgconn)` would release the lock if things go south. Please double check.

elprans replied:
I placed `unlock` behind @msullivan's `deferred_shield`

msullivan replied:
I think I had a mostly written test somewhere for `deferred_shield`; I guess I should find, finish, and commit it, if we're actually using this in the product now

fantix replied:
> The main question is if `tenant.release_pgcon(dbname, pgconn)` would release the lock if things go south. Please double check.

I think `release_pgcon()` won't release the lock? It only ends the session (thus releasing the lock) when the pgcon is not healthy, or it's simply put back to the pool with the lock being held. As session-level advisory locks cannot be released from other sessions (this is not even an error, just a PG warning) than the owning session, and there's no guarantee we can get the owning session back in the next pool.acquire() to release the lock, we may have a leak here?

```
postgres=# select pg_advisory_unlock(123);
WARNING:  you don't own a lock of type ExclusiveLock
 pg_advisory_unlock
--------------------
 f
(1 行记录)

postgres=# select pg_try_advisory_lock(123);
 pg_try_advisory_lock
----------------------
 f
(1 行记录)
```

This is probably one of the causes of the recent 6.0-rc CI failures.

fantix replied:
Addressed in #8320

1st1 commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,1305 @@
+
+            if processed == 0:
+                # No work, sleep for a bit.
+                await asyncio.sleep(10)
```
Move this to server.defines?

elprans replied:
This got replaced with a config var instead.

1st1 commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,1305 @@
+async def _ext_ai_lock(
+    pgconn: pgcon.PGConnection,
+) -> bool:
+    b = await pgconn.sql_fetch_val(b"SELECT pg_try_advisory_lock(3987734540)")
```
> 3987734540

This should also be moved to a constant

1st1 commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,1305 @@
+    elif provider.name == "builtin::mistral":
+        return await _generate_mistral_embeddings(
+            provider, model_name, inputs, shortening)
+    else:
```
What about Anthropic embeddings?

elprans replied:
Anthropic does not provide any embedding models.

elprans commented:
@scotttrinh, added you as a reviewer as I'm refactoring the auth ext tests a bit in the last few commits.

msullivan commented:
That test looks good. Do we need some more for various schema manipulations? RAG?

scotttrinh reviewed (approved):
Test changes look good!

jaclarke commented on edb/server/protocol/ai_ext.py:
```diff
@@ -0,0 +1,1346 @@
+                            b'event: content_block_start\n'
+                            + b'data: {"type": "content_block_start",'
+                            + b'"index":0,'
+                            + b'"content_block":{"type":"text","text":""}\n\n'
```
```suggestion
                            + b'"content_block":{"type":"text","text":""}}\n\n'
```
